altair/utils/transformed_data.py

```diff
@@ -0,0 +1,545 @@
+from typing import List, Optional, Tuple, Dict, Iterable, overload
+
+import pandas as pd
+
+from altair import (
+    Chart,
+    FacetChart,
+    LayerChart,
+    HConcatChart,
+    VConcatChart,
+    ConcatChart,
+    data_transformers,
+)
+from typing import Union
+from altair.utils.schemapi import Undefined
+
+Scope = Tuple[int, ...]
+FacetMapping = Dict[Tuple[str, Scope], Tuple[str, Scope]]
+
+MAGIC_CHART_NAME = "_vf_mark{}"
+
+
+@overload
+def transformed_data(
+    chart: Union[Chart, FacetChart],
+    row_limit: Optional[int] = None,
+    exclude: Optional[Iterable[str]] = None,
+) -> Optional[pd.DataFrame]:
+    ...
+
+
+@overload
+def transformed_data(
+    chart: Union[LayerChart, HConcatChart, VConcatChart, ConcatChart],
+    row_limit: Optional[int] = None,
+    exclude: Optional[Iterable[str]] = None,
+) -> List[pd.DataFrame]:
+    ...
+
+
+def transformed_data(chart, row_limit=None, exclude=None):
+    """Evaluate a Chart's transforms
+
+    Evaluate the data transforms associated with a Chart and return the
+    transformed data as one or more DataFrames
+
+    Parameters
+    ----------
+    chart : Chart, FacetChart, LayerChart, HConcatChart, VConcatChart, or ConcatChart
+        Altair chart to evaluate transforms on
+    row_limit : int (optional)
+        Maximum number of rows to return for each DataFrame. None (default) for unlimited
+    exclude : iterable of str
+        Set of the names of charts to exclude
+
+    Returns
+    -------
+    DataFrame or list of DataFrame
+        If input chart is a Chart or Facet Chart, returns a DataFrame of the transformed data
+        Otherwise, returns a list of DataFrames of the transformed data
+    """
+    try:
+        from vegafusion import runtime, get_local_tz, get_inline_datasets_for_spec  # type: ignore
+    except ImportError as err:
+        raise ImportError(
+            "transformed_data requires the vegafusion-python-embed and vegafusion packages\n"
+            "These can be installed with pip using:\n"
+            "    pip install vegafusion[embed]\n"
+            "Or with conda using:\n"
+            "    conda install -c conda-forge vegafusion-python-embed vegafusion"
+        ) from err
+
+    if isinstance(chart, Chart):
+        # Add dummy mark if None specified to satisfy Vega-Lite
+        if chart.mark == Undefined:
+            chart = chart.mark_point()
+
+    # Deep copy chart so that we can rename marks without affecting caller
+    chart = chart.copy(deep=True)
+
+    # Rename chart or subcharts with magic names that we can look up in the
+    # resulting Vega specification
+    chart_names = name_chart(chart, 0, exclude=exclude)
+
+    # Compile to Vega and extract inline DataFrames
+    with data_transformers.enable("vegafusion-inline"):
+        vega_spec = chart.to_dict(format="vega")
+        inline_datasets = get_inline_datasets_for_spec(vega_spec)
```

Comment:
I'd like to bring the `vegafusion-inline` transformer and the `get_inline_datasets_for_spec` function over to Altair eventually, but will save that for a future PR.

Reply:
Just for my own understanding, does the `vegafusion-inline` transformer use the entrypoint system to register itself or why is it available here?

Reply:
Yes, it's defined and registered in vegafusion during import: https://github.com/hex-inc/vegafusion/blob/ae2040ebf93b1a69328be9df2ec831bd8fbb6ff4/python/vegafusion/vegafusion/transformer.py#L256